### PR TITLE
Fix qualified name handling in DAML REPL

### PR DESCRIPTION
**Pull request by cocreature**
This fixes the issue reported in
https://discuss.daml.com/t/how-to-use-qualified-template-names-in-repl-queries/1329/7

We need to print names qualified otherwise, the type signature for
following lines will be ambiguous even if users qualified the name in
their input. Figuring out the right name to use in qualification is
tricky but Luckily GHC provides this already so we just have to make
sure to plug it into the right places.

changelog_begin

- [DAML REPL] Fix a bug where you got an error about a name being
  ambiguous even if you used a qualified name.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
hm, doing everything qualified seems to break things like `Party` that now reference `DA.Internal.LF.Party` :thinking:

**Comment by cocreature** [comment]
Not quite sure how to solve this, we can use GHC’s functionality for qualified imports but that falls apart as soon as you have a reexport. I guess we could track what we reference and add imports but that seems incredibly annoying.

**Comment by cocreature** [comment]
changed to a draft since this is clearly wrong in the current state

**Comment by cocreature** [comment]
Fixed it! Turns out GHC does provide functionality to print things properly qualified which makes this much much easier than I thought it would be.

You can still break things if the type isn’t in scope at all but that’s not related to this PR.

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 1]
```diff
@@ -373,10 +375,24 @@ runRepl
     -> IO ()
 runRepl importPkgs opts replClient logger ideState = do
     imports <- loadPackages importPkgs replClient ideState
+    -- Typecheck ones to get the GlobalRdrEnv
```
Comment (aherrmann-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:378]:
```suggestion
    -- Typecheck once to get the GlobalRdrEnv
```

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs [hunk 2]
```diff
@@ -373,10 +375,24 @@ runRepl
...
+    let initialLineNumber = 0
+    dflags <- liftIO $
+         hsc_dflags . hscEnv <$>
+         runAction ideState (use_ GhcSession $ lineFilePath initialLineNumber)
+    (_, tcr) <-
+        (runExceptT $ printDelayedDiagnostics $ tryTypecheck initialLineNumber $
+         T.pack (unlines $ moduleHeader dflags (importFromList imports) initialLineNumber)
```
Comment (aherrmann-da) [inline compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs:385]: Maybe factor this part out? There are another couple instances of this in the module.

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs [hunk 3]
```diff
@@ -327,6 +327,15 @@ functionalTests replClient replLogger serviceOut options ideState = describe "re
...
           , matchOutput "1"
           ]
+    , testInteraction' "collision"
+      -- Test that collisions are handled correctly if users qualify names.
+          [ input "let x = ReplTest.NameCollision \"abc\""
+          , input "x"
+          , matchOutput "NameCollision {field = \"abc\"}"
+          , input "y <- pure $ Colliding.NameCollision 42"
+          , input "y"
+          , matchOutput "NameCollision {field = 42}"
```
Comment (aherrmann-da) [inline compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:337]: Not for this PR, but I'm wondering if the outputs should be qualified too in this case.

**Review by aherrmann-da — APPROVED** [review]
Nice, thanks for looking into this!

File: compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs (hunk 2 above)
Comment (cocreature) [reply compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs]: good idea, done!

File: compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs (hunk 3 above)
Comment (cocreature) [reply compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:337]: Not sure I understand what you are suggesting. This is just the `Show` instance. We don’t have any control over this and it doesn’t depend on imports.

Comment (aherrmann-da) [reply compiler/damlc/tests/src/DA/Test/Repl/FuncTests.hs:337]: Ah, right, forgot that this is just `Show`. Nvm.

**Pull request merged**